### PR TITLE
Expose more of the Transaction type

### LINKED ISSUE #187: RUSTSEC-2021-0139: ansi_term is Unmaintained
> ansi_term is Unmaintained

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Status              | unmaintained                |
| Package             | `ansi_term`                      |
| Version             | `0.12.1`                   |
| URL                 | [https://github.com/ogham/rust-ansi-term/issues/72](https://github.com/ogham/rust-ansi-term/issues/72) |
| Date                | 2021-08-18                         |

The maintainer has adviced this crate is deprecated and will not
receive any maintenance.

The crate does not seem to have much dependencies and may or may not be ok to use as-is.

Last release seems to have been three years ago.

## Possible Alternative(s)

 The below list has not been vetted in any way and may or may not contain alternatives;

 - [anstyle](https://github.com/epage/anstyle)
 - [nu-ansi-term](https://crates.io/crates/nu-ansi-term)
 - [yansi](https://crates.io/crates/yansi)

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2021-0139.html) for additional details.

### LINKED ISSUE #303: Expose utility methods on the `Transaction` type
This issue is meant to open the discussion around exposing some utility functions defined in rust-bitcoin to the bindings.

After we merged the new [Transaction type](https://github.com/bitcoindevkit/bdk-ffi/pull/296) I started looking into what is available in it, and I think there might be value in bringing in at least 4 methods:
1. `weight()`
2. `size()`
3. `vsize()`
4. `is_absolute_timelock_satisfied()`

### PR DESCRIPTION
**Pull request by thunderbiscuit**
## Description
We've recently exposed the `Transaction` type, and I think a few methods on it would be useful. ~This is a draft PR with the first 3: `weight()`, `size()`, and `vsize()`~ _Edit: it's now much more_. I think there might be other methods we might want to expose as well. [Take a look at the docs to see them all](https://docs.rs/bitcoin/0.29.2/bitcoin/blockdata/transaction/struct.Transaction.html).

Other candidates have now been added:
1. [is_explicitly_rbf](https://docs.rs/bitcoin/0.29.2/bitcoin/blockdata/transaction/struct.Transaction.html#method.is_explicitly_rbf)
2. [is_lock_time_enabled](https://docs.rs/bitcoin/0.29.2/bitcoin/blockdata/transaction/struct.Transaction.html#method.is_lock_time_enabled)
3. [is_absolute_timelock_satisfied](https://docs.rs/bitcoin/0.29.2/bitcoin/blockdata/transaction/struct.Transaction.html#method.is_absolute_timelock_satisfied)
4. [is_coin_base](https://docs.rs/bitcoin/0.29.2/bitcoin/blockdata/transaction/struct.Transaction.html#method.is_coin_base)
5. [txid](https://docs.rs/bitcoin/0.29.2/bitcoin/blockdata/transaction/struct.Transaction.html#method.txid)

This PR is growing in size but I decided to add all 4 fields on the `Transaction` type. This is useful because it means we can now add the `transaction` field on the `TransactionDetails` type (also added in this PR).

I still have a few questions regarding all the translation between the ffi and bdk/rust types, some of the traits I had to remove (Eq and PartialEq on the TransactionDetails type) as well as the usage of `Option<Arc<T>>` vs `Arc<Option<T>>`. Will outline those tomorrow. 

Closes #303
Closes #187

## Changelog notice
```md
APIs Added
- `Transaction` type now exposes the `.weight()`, `.size()`, `.vsize()`, `is_explicitly_rbf()`, `is_lock_time_enabled()`, `is_coin_base(), and `txid()` methods [#307]

[#307]: https://github.com/bitcoindevkit/bdk-ffi/pull/307
```

#### All Submissions:
* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:
* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

**Comment by thunderbiscuit** [comment]
I added the new methods.

You can test the PR by building bdk-jvm with a custom name and publishing to local maven, and then running the following script:
```kotlin
@file:Repository("file:///~/.m2/repository/")
@file:DependsOn("org.bitcoindevkit:bdk-jvm:0.28.0-PR307")

import org.bitcoindevkit.*

val memoryDatabaseConfig = DatabaseConfig.Memory
val blockchainConfig = BlockchainConfig.Electrum(
    ElectrumConfig(
        "ssl://electrum.blockstream.info:60002",
        null,
        5u,
        null,
        100u,
        true,
    )
)

val descriptor = Descriptor("wpkh(tprv8ZgxMBicQKsPf2qfrEygW6fdYseJDDrVnDv26PH5BHdvSuG6ecCbHqLVof9yZcMoM31z9ur3tTYbSnr1WBqbGX97CbXcmp5H6qeMpyvx35B/84h/1h/0h/0/*)", Network.TESTNET)
val wallet = Wallet(descriptor, null, Network.TESTNET, memoryDatabaseConfig)
val blockchain = Blockchain(blockchainConfig)
    
wallet.sync(blockchain, null)
val balance = wallet.getBalance()
check(balance.total > 0uL) { 
     "Wallet balance is 0, please add funds!"
}

@OptIn(ExperimentalUnsignedTypes::class)
val faucetAddress = "tb1ql7w62elx9ucw4pj5lgw4l028hmuw80sndtntxt"
val (psbt, txDetails) = TxBuilder()
        .addRecipient(Address(faucetAddress).scriptPubkey(), 10000u)
        .feeRate(2.0f)
        .enableRbf()
        .finish(wallet)

val tx = psbt.extractTx()
println("Transaction info is:")
println("Coinbase: ${tx.isCoinBase()}")
println("Explicitly signals RBF: ${tx.isExplicitlyRbf()}")
println("Locktime is enabled: ${tx.isLockTimeEnabled()}")
println("Txid: ${tx.txid()}")
println("Weight: ${tx.weight()}")
println("Size: ${tx.size()}")
println("Virtual size: ${tx.vsize()}")
```

**Comment by thunderbiscuit** [comment]
@notmandatory this is now ready for review.

File: bdk-ffi/src/lib.rs [hunk 1]
```diff
@@ -189,24 +200,22 @@ pub struct LocalUtxo {
     is_spent: bool,
 }
 
-// This trait is used to convert the bdk TxOut type with field `script_pubkey: Script`
-// into the bdk-ffi TxOut type which has a field `address: String` instead
 trait NetworkLocalUtxo {
```
[the remark this answers is not in the document]
Comment (thunderbiscuit) [reply bdk-ffi/src/lib.rs:214]: Removed indeed.

File: bdk-ffi/src/lib.rs [hunk 2]
```diff
@@ -96,8 +96,9 @@ impl From<AddressIndex> for BdkAddressIndex {
 }
 
 /// A wallet transaction
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[derive(Debug, Clone, Default)]
 pub struct TransactionDetails {
+    pub transaction: Option<Arc<Transaction>>,
```
[the remark this answers is not in the document]
Comment (thunderbiscuit) [reply bdk-ffi/src/lib.rs:101]: Done.

File: bdk-ffi/src/lib.rs [hunk 3]
```diff
@@ -189,27 +206,21 @@ pub struct LocalUtxo {
...
+                vout: local_utxo.outpoint.vout,
             },
             txout: TxOut {
-                value: x.txout.value,
-                address: BdkAddress::from_script(&x.txout.script_pubkey, network)
-                    .unwrap()
-                    .to_string(),
+                value: local_utxo.txout.value,
+                script_pubkey: Arc::new(Script {
+                    script: local_utxo.txout.script_pubkey.clone(),
```
Comment (notmandatory) [inline bdk-ffi/src/lib.rs:219]: nit, can we do this without a clone?

File: bdk-ffi/src/lib.rs [hunk 4]
```diff
@@ -238,8 +249,32 @@ impl fmt::Debug for ProgressHolder {
...
 
+#[derive(Debug, Clone)]
+pub struct TxIn {
+    pub previous_output: OutPoint,
+    pub script_sig: Arc<Script>,
+    pub sequence: u32,
+    pub witness: Vec<Vec<u8>>,
+}
+
+impl From<&bdk::bitcoin::blockdata::transaction::TxIn> for TxIn {
```
Comment (notmandatory) [inline bdk-ffi/src/lib.rs:260]: nit:
```suggestion
use bdk::bitcoin::blockdata::transaction::TxIn as BdkTxIn;
...
impl From<&BdkTxIn> for TxIn {

File: bdk-ffi/src/psbt.rs [hunk 5]
```diff
@@ -34,7 +34,9 @@ impl PartiallySignedTransaction {
     /// Return the transaction.
     pub(crate) fn extract_tx(&self) -> Arc<Transaction> {
         let tx = self.internal.lock().unwrap().clone().extract_tx();
-        Arc::new(Transaction { internal: tx })
+        Arc::new(tx.into())
+        // let buffer: Vec<u8> = serialize(&tx);
```
Comment (notmandatory) [inline bdk-ffi/src/psbt.rs:38]: remove commented out lines.

**Review by notmandatory — APPROVED** [review]
ACK cba69e681a2116c98895d021723617f71db1d0ca

**Pull request merged**